src/vm.rs

```diff
@@ -1178,3 +821,428 @@ impl<'a, E: UserDefinedError, I: InstructionMeter> EbpfVm<'a, E, I> {
         }
     }
 }
+
+/// EbpfInterpreter is the in-flight state of the eBPF interpreter.
+///
+/// It borrows the VM and instruction meter for lifetime 'b, which is outlived by 'a.
+pub struct EbpfInterpreter<'a, 'b, E: UserDefinedError, I: InstructionMeter> {
+    vm: &'b mut EbpfVm<'a, E, I>,
+    instruction_meter: &'b mut I,
+
+    /// General purpose registers.
+    reg: [u64; 11],
+    /// Program counter of the instruction that was last executed.
+    next_pc: usize,
+
+    /// Number of remaining instructions allowed.
+    remaining_insn_count: u64,
+    /// Number of instructions initially allowed.
+    initial_insn_count: u64,
+    /// Number of instructions executed so far.
+    total_insn_count: u64,
+
+    instruction_meter_enabled: bool,
+    instruction_tracing_enabled: bool,
+    dynamic_stack_frames: bool,
+}
+
+impl<'a, 'b, E: UserDefinedError, I: InstructionMeter> EbpfInterpreter<'a, 'b, E, I> {
+    fn run(&mut self) -> ProgramResult<E> {
+        loop {
+            self.check_execution_overrun()?;
+            if let Some(r0) = self.step()? {
+                return Ok(r0);
+            }
+        }
+    }
+
+    fn check_execution_overrun(&self) -> Result<(), EbpfError<E>> {
```

Comment:
This function should probably merged into `step`. There is no valid reason to execute `step` without checking for an overrun.